README.md
# bigrest

[![NPM Version][npm-image]][npm-url]
[![NPM Downloads][downloads-image]][downloads-url]

Smart web and restful framework for NodeJS. current core is [**express**](http://expressjs.com).

## Installation

```sh
$ npm install bigrest
```

## Usage

```js
var bigrest = require('bigrest')
```

## API

### listen(port, opts)

Binds and listens for connections on the port.

``` javascript
var path = require('path');
var bigrest = require('../index');

var http = bigrest.listen(18080, {
    basepath: __dirname,
    statics: [{
        urlpath: '/files',
        filepath: path.join(__dirname, 'files')
    }],
    compression: {
        threshold: 16 * 1024
    }
});

```

> XArray(string) was a Array(string) or a Array splited by ",".  
> XArray(*object*) was a Array(*object*) or a Object(*object*) ",".

### Parameter: opts

The options for bigrest framework.

|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|debug|boolean|false|debug mode switch|
|etag|boolean|false|switch on **ETAG** generation|
|https|Object|undefined|options for HTTPS protocol|
|compression|boolean or Object|false|switch or options for compression|
|basepath|string|process.cwd()|the base path|
|services|XArray(string)|services|the **handlers and routers** path|
|rootwork|function|undefined|the all METHOD handler for '/'|
|r404work|function|undefined|the all METHOD handler while response 404|
|visitor|function|undefiend|the pre processor the all handler|
|viewer|Object(viewer)|undeifned|set html template engine|
|limits|Object(limits)|undeifned|the limits for request parse|
|statics|XArray(static)|undefiend|set static file engine|
|middlewares|Array(function)|[]|the middlewares for express|

> https documents [https.createServer](https://nodejs.org/api/https.html#https_https_createserver_options_requestlistener)  
> compression as **options** documents [compression](https://github.com/expressjs/compression#compressionoptions)

#### Object(viewer)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|filepath|string|undefined|the template source file path|
|render|string or function|undefiend|the template engine|
|cache|boolean|opposite with opts.debug|system cache switch|

```javascript
{
    render: swig.renderFile,
    filepath: path.join(__dirname, "server", "views")
}
```

#### Object(limits)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|bodySize|number|4mb|the request body size|
|uploadSize|number|50mb|the maximum size for file upload|

```javascript
{
    bodySize: 102400,
    uploadSize: 1024000
}
```

#### Object(static)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|urlpath|string|undefined|the url path|
|filepath|string|undefined|the static file's file path|
|options|Object|undefined|the static options|

> *options* means in the [express.static](http://expressjs.com/en/4x/api.html#express.static)

```javascript
{
    urlpath: '/static',
    filepath: path.join(__dirname, "static")
}
```

## Directory

```sh
|--{basepath}
|   |--commons
|   |   |--test.js
|   |--methods
|   |   |--test.js
|   |--services
|   |   |--test-router.json
|   |   |--test.js
|--server.js
```

### Global variables

|*Name*|*Path*|*Export Variables*|*Export methods*|
|---|---|---|---|
|brcx|commons/|*UpperCase named, number & string*|all|
|brmx|method/|*UpperCase named, number & string*|lowercase named|

### Router And Processor

> services/test-router.json

``` json
{
    "processors": [
        {
            "url": "/benchmark",
            "method": "GET",
            "processor": "benchmark_empty",
            "parameters": []
        }
    ]
}
```

> services/test.js

``` javascript
exports.benchmark_empty = function(req, res) {
    brmx.benchmark(function() {
        res.send('OK');
    });
};
```

> methods/test.js

``` javascript
exports.benchmark = function(callback) {
    callback(null);
};
```

#### Router File Format

it has three style, Object(container), Array(group), Object(group).

##### Object(container)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|parameters|Array(Parameger)|[]|the common parameters|
|interceptor|XArray(string)|[]|the common interceptor names|
|failure|stirng|undefined|the default failure processor name|
|groups|Array(group)|**required**|the router group|

##### Object(group)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|parameters|Array(Parameger)|[]|the common parameters|
|interceptor|XArray(string)|[]|the interceptor names|
|failure|stirng|undefined|the failure processor name|
|processors|Array(processor)|**required**|the router parameters|

##### Object(processor)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|method|XArray(string)|**required**|METHOD|
|url|XArray(string)|**required**|URL|
|parameters|Array(parameger)|[]|the common parameters|
|failure|stirng|undefined|the failure processor name|
|processor|string|**required**|the processor name|
|workdata|Object|undefined|attach to **req**|
|workparam|Object|undefined|attach to **req**|

##### Object(parameter)
|*Key*|*Type*|*Default*|*Description*|
|---|---|---|---|
|name|string|**required**|the name|
|empty|boolean|false|accept empty value|,
|default|stringOrNumber|undefined|the default value when empty|
|candidate|string|undefined|accept empty when **has** another prameter|
|match|string|undefined|accept value when **equal** another prameter|
|trim|boolean|false|trim the value string|
|digit|boolean|false|accept digit number value|
|length|Array(number)[min,max]|undefined|accept vlaue **length** **IN** range|
|length_min|number|1|accept value minimal **length**|
|length_max|number|undefined|accept value maximal **length**|
|word|boolean|false|accept value match **Word**|
|regexp|string|false|accept value match **regexp**|
|values|Array(stringOrNumber)|undefined|validate value **IN** Array|
|range|Array(number)[min,max]|undefined|accept vlaue **IN** range|
|range_min|number|undefined|accept minimal value|
|range_max|number|undefined|accept maximal value|

### req.files Information

It's array of file, And each file contains the following information:

|*Key*|*Description*|
|---|---|
|`name` | Field name specified in the form |
|`type` | Mime type of the file |
|`size` | Size of the file in bytes |
|`path` | The full path to the temp uploaded file |
|`originalFilename` | Name of the file on the user's computer |

## Examples

[demo](https://github.com/vietor/bigrest/tree/master/demo)  
[bigbbs](https://github.com/vietor/bigbbs)

## License

[MIT](LICENSE)

[npm-image]: https://img.shields.io/npm/v/bigrest.svg
[npm-url]: https://npmjs.org/package/bigrest
[downloads-image]: https://img.shields.io/npm/dm/bigrest.svg
[downloads-url]: https://npmjs.org/package/bigrest


HISTORY.md
2.0.0
========

* Relace connect-multipart to multer
* Remove opts.static support
* Modify default bodySize limit to 4mb
* Upgrade some deps

1.5.8
========

* Use 'strict mode' for 'application/json'
* Fix 'req.query' merge the 'req.body' & 'req.params'

1.5.7
========

* Fix __bigrest_params usage since 1.5.5

1.5.6
========

* Fix req.param() implementation
* deps: express@~4.15.3
* deps: body-parser@~1.17.2

1.5.5
========

* Refactory code for parameter validator
* Performance the req.param() implementation

1.5.4
========

* Fix intercepter result check

1.5.3
========

* Fix opts.statics usage

1.5.2
========

* Ignore error on remove the upload temp file.
* Add more compatible for newest NodeJS.
* deps: express@~4.15.2
* deps: body-parser@~1.17.1

1.5.1
========

* Remove session support, use opts.middlewares for support it.
* Add more check for interceptor
* deps: lodash@~4.17.4
* deps: express@~4.14.1
* deps: body-parser@~1.16.1

1.5.0
========

* Add options.cookieSession support
* Remove options.session.store 'function' support
* Rename options.session.storeNeedReady to storeStatusCheck

1.4.10
========

  * Add 'r404work' for options.

1.4.9
========

  * Refactory code for **error handle**

1.4.8
========

  * Add more check for **error handle**
  * deps: lodash@~4.17.2

1.4.7
========

  * Modify some options attribute support
  * deps: lodash@~4.16.4

1.4.6
========

  * Add length[min,max], range_min, range_max for parameter check
  * Add lodash usage
  * Refactory parameter test
  * deps: express-session@~1.14.1

1.4.5
========

  * deps: body-parser@~1.15.2
  * deps: express@~4.14.0
  * deps: express-session@~1.14.0

1.4.4
========

  * Add etag switch for options
  * Add compression options support
  * deps: compression@~1.6.2
  * deps: cookie-parser@~1.4.3

1.4.3
========

  * Refactory some code
  * Add more HTTP code support
  * deps: body-parser@~1.15.1

1.4.2
========

  * Add session initialize check

1.4.0
========

  * New base release

1.3.12
========

  * Add attribute cache for viewer
  * Add document for routet file

1.3.11
========

  * Add compression support

1.3.10
========

  * Add size limit in opts.limits
  * deps: express@~4.13.4
  * deps: body-parser@~1.15.0

1.3.9
========

  * Add array support with opts.static

1.3.8
========

  * Add more attribute (failure, match) for router

1.3.7
========

  * Add HTTP-400 as default failure for services
  * deps: cookie-parser@~1.4.1
  * deps: express-session@~1.13.0

1.3.6
========

  * Split template files
  * Remove multiple error middlewares for express

1.3.5
========

  * Fix crash when error catch in express

1.3.4
========

  * Hotfix for 1.3.3

1.3.3
========

  * Performance and clean code
  * deps: body-parser@~1.14.2

1.3.2
========

  * Upgrade upstream
  * Throw exception when simulator

1.3.1
========

  * Clean option
    - Remove nocache
    - Remove commons, methods
  * Dump error stack just when debug enabled

1.3.0
========

  * First public version, base from 1.2.8


lib/bigrest.js
"use strict";

var fs = require('fs');
var path = require('path');
var xutil = require('./xutil');
var commons = require('./commons');
var methods = require('./methods');
var services = require('./services');

var template = xutil.readFileText('code.html');
var codetexts = (function() {
    var texts = {};
    var jsontext = xutil.readFileText('codes.json');
    xutil.eachObject(JSON.parse(jsontext), function(message, code) {
        texts[code] = xutil.renderTemplateText(template, {
            code: code,
            message: message
        });
    });
    return texts;
})();

function ishttperr(err) {
    return err && (err.statusCode && err.status) && (err.statusCode == err.status) && (err.status >= 100 && err.status < 600);
}

function textcode(code) {
    return codetexts[code] || xutil.renderTemplateText(template, {
        code: code,
        message: 'Unexpected Error'
    });
}

function options(opts) {
    if (!opts)
        opts = {};
    if (!opts.limits)
        opts.limits = {};
    if (!opts.basepath)
        opts.basepath = process.cwd();
    if (!opts.services)
        opts.services = "services";
    opts._privates = {
        default_failure: function(req, res, name, type) {
            res.status(400).send(opts.debug ? ("[" + name + ":" + type + "] " + req.path) : textcode(400));
        }
    };
    return opts;
}

function present(opts) {
    var _brcommons_ = {};
    commons.setup(_brcommons_, path.join(opts.basepath, "commons"));
    global.brcx = global.brcommons = _brcommons_;
    var _brmethods_ = {};
    methods.setup(_brmethods_, path.join(opts.basepath, "methods"));
    global.brmx = global.brmethods = _brmethods_;
}

function weblike(opts) {
    var express = require('express');
    var bodyParser = require('body-parser');
    var cookieParser = require('cookie-parser');
    var multer = require('multer');
    var compression = require('compression');

    present(opts);

    var app = express();
    if (!opts.etag)
        app.disable('etag');
    app.disable('x-powered-by');
    app.enable('trust proxy');
    if (opts.compression) {
        var compressionOptions = {};
        if (opts.compression !== true)
            compressionOptions = opts.compression;
        app.use(compression(compressionOptions));
    }
    if (opts.viewer) {
        var renderType = typeof opts.viewer.render;
        if (["string", "function"].indexOf(renderType) == -1)
            throw new Error("Bad option: viewer.render");
        if (typeof opts.viewer.filepath != 'string')
            throw new Error("Bad option: viewer.filepath");
        app.set('views', opts.viewer.filepath);
        if (renderType == 'string')
            app.set('view engine', opts.viewer.render);
        else {
            app.set('view engine', 'html');
            app.engine('html', opts.viewer.render);
        }
        app.set('view cache', xutil.optBoolean(opts.viewer.cache, !opts.debug));
    }
    var bodySize = opts.limits.bodySize || (4 * 1024 * 1024);
    app.use(bodyParser.json({
        strict: true,
        limit: bodySize
    }));
    app.use(bodyParser.urlencoded({
        extended: false,
        limit: bodySize
    }));
    app.use(multer({
        storage: multer.diskStorage({}),
        limits: {
            fieldNameSize: 256,
            fieldSize: bodySize,
            fileSize: opts.limits.uploadSize || (50 * 1024 * 1024)
        }
    }).any());
    app.use(function(req, res, next) {
        if (req.files) {
            req.files = xutil.mapArray(req.files || [], function(file) {
                return {
                    name: file.fieldname,
                    type: file.mimetype,
                    size: file.size,
                    path: file.path,
                    originalFilename: file.originalname
                };
            });
        }
        next();
    });
    app.use(cookieParser());
    xutil.eachArray(opts.middlewares, function(middleware) {
        app.use(middleware);
    });
    app.use(function(req, res, next) {
        res.setHeader('Server', 'nginx');
        req.takeParamsRef = function() {
            if (!req.__bigrest_params)
                req.__bigrest_params = xutil.objExtend({}, req.query || {}, req.body || {}, req.params || {});
            return req.__bigrest_params;
        };
        req.param = function(name, defaultValue) {
            var value = req.takeParamsRef()[name];
            if (xutil.isNullOrUndefined(value))
                value = defaultValue;
            return value;
        };
        var fks = req.files || [];
        if (fks.length > 0) {
            res.on('finish', function() {
                xutil.eachArray(fks, function(k) {
                    try {
                        fs.unlinkSync(k.path);
                    } catch (e) {}
                });
            });
        }
        next();
    });
    if (opts.rootwork) {
        app.all("/", opts.rootwork);
    }
    xutil.eachArray(xutil.parseArray(opts.services), function(dirname) {
        services.setup(app, path.join(opts.basepath, dirname), xutil.objExtend({
            visitor: opts.visitor
        }, opts._privates));
    });
    if (opts.statics) {
        xutil.eachArray(xutil.parseArray(opts.statics), function(row) {
            if (typeof row.urlpath != 'string')
                throw new Error("Bad option: static.urlpath");
            if (typeof row.filepath != 'string')
                throw new Error("Bad option: static.filepath");
            app.use(row.urlpath, express.static(row.filepath, row.options || {}));
        });
    }
    if (opts.r404work) {
        app.use(opts.r404work);
    }
    app.use(function(req, res) {
        res.status(404).send(opts.debug ? (req.method + " " + req.path) : textcode(404));
    });
    app.use(function(err, req, res, next) {
        // Add error entry lock
        if (req.__bigrest_error_lock)
            next(err);
        else {
            req.__bigrest_error_lock = true;

            if (res.headersSent)
                next(err);
            else {
                var errors = [];
                var recatch = false;
                try {
                    var statusCode = 500;
                    if (!ishttperr(err))
                        errors.push(err);
                    else
                        statusCode = err.statusCode;
                    res.status(statusCode).send(opts.debug ? (err.stack || err.message) : textcode(statusCode));
                } catch (e) {
                    errors.push(e);
                    recatch = true;
                }
                if (opts.catcher && errors.length > 0) {
                    try {
                        opts.catcher(errors, req, res);
                    } catch (e) {
                        recatch = true;
                    }
                }
                // Use the default handler when error again
                if (recatch) {
                    next(err);
                }
            }
        }
    });
    return app;
}

exports.present = function(opts) {
    present(options(opts));
};

exports.listen = function(port, opts) {
    opts = options(opts);
    var http = require('http');
    var https = require('https');

    var server;
    var webapp = weblike(opts);
    if (!opts.https)
        server = http.createServer(webapp);
    else
        server = https.createServer(opts.https, webapp);
    var new_arguments = [port];
    for (var i = 2; i < arguments.length; ++i)
        new_arguments.push(arguments[i]);
    server.listen.apply(server, new_arguments);
};

exports.simulator = function(opts) {
    opts = xutil.objExtend({
        debug: true
    }, options(opts));

    var http = require('http');
    var qs = require('querystring');

    var webapp = weblike(opts);
    return {
        request: function(method, url, params, callback) {
            var fakereq = (function() {
                var qss = qs.stringify(params);
                var req = new http.IncomingMessage();
                req.url = url;
                req.method = method.toUpperCase();
                if (req.method == 'GET') {
                    if (req.url.indexOf('?') == -1)
                        req.url += '?';
                    else
                        req.url += '&';
                    req.url += qss;
                } else {
                    var readed = false;
                    var buffer = new Buffer(qss);
                    req.headers['content-type'] = 'application/x-www-form-urlencoded';
                    req.headers['content-length'] = buffer.length;
                    req._read = function() {
                        if (readed)
                            req.push(null);
                        else {
                            readed = true;
                            req.push(buffer);
                        }
                    };
                }
                req.ip = xutil.getFakeIp();
                req.ips = [req.ip];
                return req;
            })();
            var fakeres = (function() {
                var resend = false;
                var res = new http.ServerResponse(fakereq);
                res.send = function(data) {
                    if (resend)
                        return;
                    resend = true;
                    process.nextTick(function() {
                        callback(res.statusCode, data);
                    });
                };
                return res;
            })();
            webapp(fakereq, fakeres);
        }
    };
};
